File: backends/micronaut-java/src/main/java/dev/stackverse/backend/AdminController.java
package dev.stackverse.backend;

import io.micronaut.http.HttpRequest;
import io.micronaut.http.HttpResponse;
import io.micronaut.http.MutableHttpResponse;
import io.micronaut.http.annotation.Body;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.PathVariable;
import io.micronaut.http.annotation.Put;
import io.micronaut.scheduling.TaskExecutors;
import io.micronaut.scheduling.annotation.ExecuteOn;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.Connection;
import java.sql.SQLException;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import tools.jackson.core.type.TypeReference;
import tools.jackson.databind.ObjectMapper;

@Controller
@ExecuteOn(TaskExecutors.BLOCKING)
final class AdminController {
    private static final Logger LOG = LoggerFactory.getLogger(AdminController.class);

    private final Database db;
    private final SecuritySupport security;
    private final AccountService accounts;
    private final AuditService audit;
    private final ObjectMapper mapper;

    AdminController(Database db, SecuritySupport security, AccountService accounts, AuditService audit, ObjectMapper mapper) {
        this.db = db;
        this.security = security;
        this.accounts = accounts;
        this.audit = audit;
        this.mapper = mapper;
    }

    @Get("/api/v1/admin/users")
    PageResponse<AccountResponse> users(HttpRequest<?> request) {
        security.requireRole(request, "admin");
        int page = WebSupport.page(request);
        int size = WebSupport.size(request);
        String q = request.getParameters().getFirst("q").orElse("");
        if (WebSupport.length(q) > 100) {
            throw Problems.badRequest("q is too long");
        }
        String status = request.getParameters().getFirst("status").orElse("");
        if (!status.isBlank() && !List.of(Models.USER_ACTIVE, Models.USER_BLOCKED).contains(status)) {
            throw Problems.badRequest("status must be one of: active, blocked");
        }
        List<AccountResponse> items = accounts.search(q, status, page, size).stream().map(this::accountResponse).toList();
        long total = accounts.countSearch(q, status);
        return WebSupport.pageResponse(items, page, size, total);
    }

    @Get("/api/v1/admin/users/{username}")
    AccountResponse user(HttpRequest<?> request, @PathVariable String username) {
        security.requireRole(request, "admin");
        return accountResponse(accounts.get(username));
    }

    @Put("/api/v1/admin/users/{username}/status")
    AccountResponse setUserStatus(HttpRequest<?> request, @PathVariable String username, @Body UserStatusInput body) {
        Identity actor = security.requireRole(request, "admin");
        if (body == null || body.status() == null) {
            throw Problems.badRequest("status is required");
        }
        if (!List.of(Models.USER_ACTIVE, Models.USER_BLOCKED).contains(body.status())) {
            throw Problems.badRequest("status must be one of: active, blocked");
        }
        accounts.get(username);
        String reason = body.reason() == null ? null : body.reason().trim();
        boolean blocking = Models.USER_BLOCKED.equals(body.status());
        if (blocking) {
            Validator validator = new Validator();
            validator.check(reason != null && !reason.isBlank(), "reason", "validation.block.reason.required");
            validator.check(WebSupport.length(reason) <= 1000, "reason", "validation.block.reason.too-long");
            validator.throwIfInvalid();
            if (username.equals(actor.username())) {
                throw Problems.conflict("Admins cannot block themselves.");
            }
        }
        String status = blocking ? Models.USER_BLOCKED : Models.USER_ACTIVE;
        String auditAction = blocking ? "user.blocked" : "user.unblocked";
        String event = blocking ? "user_blocked" : "user_unblocked";
        String message = blocking ? "User account blocked" : "User account unblocked";
        db.inTx(connection -> {
            try {
                accounts.setStatus(connection, username, status, blocking ? reason : null);
                audit.record(connection, actor.username(), auditAction, "user", username,
                        blocking ? Map.of("reason", reason) : null);
            } catch (SQLException ex) {
                throw new IllegalStateException(ex);
            }
            return null;
        });
        EventLog.info(LOG, event, "success", message,
                Map.of("actor", actor.username(), "resource_type", "user", "resource_id", username));
        return accountResponse(accounts.get(username));
    }

    @Get("/api/v1/admin/audit-log")
    PageResponse<AuditResponse> auditLog(HttpRequest<?> request) {
        security.requireRole(request, "admin");
        int page = WebSupport.page(request);
        int size = WebSupport.size(request);
        String actor = request.getParameters().getFirst("actor").orElse("");
        String action = request.getParameters().getFirst("action").orElse("");
        String targetType = request.getParameters().getFirst("targetType").orElse("");
        String targetId = request.getParameters().getFirst("targetId").orElse("");
        Instant from = instantParam(request, "from");
        Instant to = instantParam(request, "to");
        String where = """
                where (? = '' or actor = ?)
                  and (? = '' or action = ?)
                  and (? = '' or target_type = ?)
                  and (? = '' or target_id = ?)
                  and (cast(? as timestamptz) is null or created_at >= ?)
                  and (cast(? as timestamptz) is null or created_at <= ?)
                """;
        List<Object> filterArgs = auditFilterArgs(actor, action, targetType, targetId, from, to);
        long total = db.scalarLong("select count(*) from audit_entries " + where, filterArgs.toArray());
        List<Object> queryArgs = new ArrayList<>(filterArgs);
        queryArgs.add(size);
        queryArgs.add(WebSupport.offset(page, size));
        List<AuditResponse> items = db.query("""
                select id, actor, action, target_type, target_id, detail, created_at
                from audit_entries %s
                order by created_at desc, id desc
                limit ? offset ?
                """.formatted(where), rs -> new AuditResponse(
                rs.getObject("id", UUID.class),
                rs.getString("actor"),
                rs.getString("action"),
                rs.getString("target_type"),
                rs.getString("target_id"),
                json(rs.getString("detail")),
                rs.getTimestamp("created_at").toInstant()
        ), queryArgs.toArray());
        return WebSupport.pageResponse(items, page, size, total);
    }

    @Get("/api/v1/admin/stats")
    MutableHttpResponse<?> stats(HttpRequest<?> request) {
        security.requireRole(request, "moderator");
        StatsTotals totals = db.one("""
                select
                    (select count(*) from user_accounts) as users,
                    (select count(*) from bookmarks) as bookmarks,
                    (select count(*) from bookmarks where visibility = 'public') as public_bookmarks,
                    (select count(*) from bookmarks where status = 'hidden') as hidden_bookmarks,
                    (select count(*) from reports where status = 'open') as open_reports
                """, rs -> new StatsTotals(rs.getLong("users"), rs.getLong("bookmarks"),
                rs.getLong("public_bookmarks"), rs.getLong("hidden_bookmarks"), rs.getLong("open_reports")));
        LocalDate today = LocalDate.now(ZoneOffset.UTC);
        LocalDate from = today.minusDays(29);
        Map<String, Long> bookmarksCreated = countsByDay("bookmarks", "created_at", from);
        Map<String, Long> activeUsers = countsByDay("user_accounts", "last_seen", from);
        List<DailyStat> daily = new ArrayList<>();
        for (int i = 0; i < 30; i++) {
            String date = from.plusDays(i).toString();
            daily.add(new DailyStat(date, bookmarksCreated.getOrDefault(date, 0L), activeUsers.getOrDefault(date, 0L)));
        }
        List<TagCount> tags = db.query("""
                select tag, count(*) as count
                from bookmarks b cross join unnest(b.tags) as tag
                group by tag
                order by count(*) desc, tag
                limit 10
                """, rs -> new TagCount(rs.getString("tag"), rs.getLong("count")));
        return WebSupport.etag(mapper, request, new AdminStats(totals, daily, tags), null);
    }

    private AccountResponse accountResponse(Account account) {
        return new AccountResponse(account.username(), account.firstSeen(), account.lastSeen(), account.status(),
                account.blockedReason(), account.bookmarkCount());
    }

    private Instant instantParam(HttpRequest<?> request, String name) {
        String raw = request.getParameters().getFirst(name).orElse("");
        if (raw.isBlank()) {
            return null;
        }
        try {
            return Instant.parse(raw);
        } catch (RuntimeException ex) {
            throw Problems.badRequest(name + " must be an RFC 3339 timestamp");
        }
    }

    private Object json(String raw) {
        if (raw == null) {
            return null;
        }
        try {
            return mapper.readValue(raw, new TypeReference<Map<String, Object>>() {
            });
        } catch (Exception ex) {
            return raw;
        }
    }

    private List<Object> auditFilterArgs(String actor, String action, String targetType, String targetId, Instant from, Instant to) {
        List<Object> args = new ArrayList<>();
        args.add(actor);
        args.add(actor);
        args.add(action);
        args.add(action);
        args.add(targetType);
        args.add(targetType);
        args.add(targetId);
        args.add(targetId);
        args.add(from);
        args.add(from);
        args.add(to);
        args.add(to);
        return args;
    }

    private Map<String, Long> countsByDay(String table, String column, LocalDate from) {
        Map<String, Long> counts = new HashMap<>();
        db.query("""
                select (%s at time zone 'UTC')::date as day, count(*) as count
                from %s
                where %s >= ?
                group by day
                """.formatted(column, table, column), rs -> {
            counts.put(rs.getObject("day", LocalDate.class).toString(), rs.getLong("count"));
            return null;
        }, from.atStartOfDay().toInstant(ZoneOffset.UTC));
        return counts;
    }
}

File: backends/ruby-rails/app/controllers/admin_controller.rb
class AdminController < ApplicationController
  def reports
    require_role!("moderator")
    page, size = query_params.page_and_size
    status = Stackverse::InputValidation.validate_report_status(query_params.single("status")) || "open"
    rows = Stackverse::Sql.query(<<~SQL.squish)
      select * from reports where status = #{Stackverse::Sql.quote(status)}
      order by created_at asc, id asc
      limit #{size} offset #{page * size}
    SQL
    total = Stackverse::Sql.one("select count(*)::int as count from reports where status = #{Stackverse::Sql.quote(status)}")["count"].to_i
    render_json Stackverse::Sql.page(rows, page, size, total, Stackverse::Serializers.method(:report))
  end

  def resolve_report
    caller = require_role!("moderator")
    report_id = Stackverse::InputValidation.parse_uuid(params[:id])
    target, note = Stackverse::InputValidation.validate_resolution(body_hash)
    row = nil
    ActiveRecord::Base.transaction do
      if target == "actioned"
        bookmark_ref = Stackverse::Sql.one("select bookmark_id from reports where id = #{Stackverse::Sql.quote(report_id)}::uuid")
        Stackverse::Errors.not_found unless bookmark_ref
        Stackverse::Sql.connection.execute("select id from bookmarks where id = #{Stackverse::Sql.quote(bookmark_ref["bookmark_id"])}::uuid for update")
      end

      report = Stackverse::Sql.one("select * from reports where id = #{Stackverse::Sql.quote(report_id)}::uuid for update")
      Stackverse::Errors.not_found unless report

      if target == "open"
        row = reopen_report!(report, caller)
      else
        row = resolve_one!(report, target, caller.username, note, false)
        if target == "actioned"
          hide_bookmark!(report["bookmark_id"].to_s, caller.username, note)
          siblings = Stackverse::Sql.query(<<~SQL.squish)
            select * from reports
            where bookmark_id = #{Stackverse::Sql.quote(report["bookmark_id"])}::uuid
              and status = 'open'
              and id <> #{Stackverse::Sql.quote(report_id)}::uuid
            order by id asc
            for update
          SQL
          siblings.each { |sibling| resolve_one!(sibling, "actioned", caller.username, note, true) }
        end
      end
    end
    render_json Stackverse::Serializers.report(row)
  rescue ActiveRecord::RecordNotUnique
    Stackverse::Errors.conflict("The reporter already has another open report on this bookmark.")
  end

  def set_bookmark_status
    caller = require_role!("moderator")
    bookmark_id = Stackverse::InputValidation.parse_uuid(params[:id])
    status, note = Stackverse::InputValidation.validate_bookmark_status(body_hash)
    row = nil
    existing = nil
    ActiveRecord::Base.transaction do
      existing = Stackverse::Sql.one("select * from bookmarks where id = #{Stackverse::Sql.quote(bookmark_id)}::uuid for update")
      Stackverse::Errors.not_found unless existing
      row = Stackverse::Sql.one(<<~SQL.squish)
        update bookmarks
        set status = #{Stackverse::Sql.quote(status)},
            updated_at = #{Stackverse::Sql.quote(Stackverse::Clock.now)}
        where id = #{Stackverse::Sql.quote(bookmark_id)}::uuid
        returning *
      SQL
      Stackverse::Audit.record(
        caller.username,
        "bookmark.status-changed",
        "bookmark",
        bookmark_id,
        { from: existing["status"], to: status, note: note }
      )
    end
    Stackverse::EventLog.info(
      "bookmark_status_changed",
      "success",
      "Bookmark moderation status changed",
      actor: caller.username,
      resource_type: "bookmark",
      resource_id: bookmark_id,
      from: existing["status"],
      to: status
    )
    render_json Stackverse::Serializers.bookmark(row)
  end

  def users
    require_role!("admin")
    page, size = query_params.page_and_size
    q = query_params.single("q")
    Stackverse::InputValidation.max_length(q, 100, "q")
    status = query_params.single("status")
    Stackverse::Errors.bad_request("unknown status: #{status}") if status && !%w[active blocked].include?(status)

    conditions = [ "true" ]
    conditions << "u.username ilike #{Stackverse::Sql.quote(Stackverse::Sql.like_pattern(q))} escape E'\\\\'" if q.present?
    conditions << "u.status = #{Stackverse::Sql.quote(status)}" if status
    where_sql = conditions.join(" and ")
    rows = Stackverse::Sql.query(<<~SQL.squish)
      #{with_bookmark_count} where #{where_sql}
      order by u.last_seen desc, u.username asc
      limit #{size} offset #{page * size}
    SQL
    total = Stackverse::Sql.one("select count(*)::int as count from user_accounts u where #{where_sql}")["count"].to_i
    render_json Stackverse::Sql.page(rows, page, size, total, Stackverse::Serializers.method(:user_account))
  end

  def user
    require_role!("admin")
    row = find_account(params[:username])
    Stackverse::Errors.not_found unless row
    render_json Stackverse::Serializers.user_account(row)
  end

  def set_user_status
    caller = require_role!("admin")
    username = params[:username]
    status, reason = Stackverse::InputValidation.validate_user_status(body_hash, username, caller.username)
    ActiveRecord::Base.transaction do
      existing = Stackverse::Sql.one("select username from user_accounts where username = #{Stackverse::Sql.quote(username)} for update")
      Stackverse::Errors.not_found unless existing
      if status == "blocked"
        Stackverse::Sql.connection.execute(<<~SQL.squish)
          update user_accounts
          set status = 'blocked', blocked_reason = #{Stackverse::Sql.quote(reason)}
          where username = #{Stackverse::Sql.quote(username)}
        SQL
        Stackverse::Audit.record(caller.username, "user.blocked", "user", username, { reason: reason })
      else
        Stackverse::Sql.connection.execute(<<~SQL.squish)
          update user_accounts
          set status = 'active', blocked_reason = null
          where username = #{Stackverse::Sql.quote(username)}
        SQL
        Stackverse::Audit.record(caller.username, "user.unblocked", "user", username)
      end
    end
    Stackverse::EventLog.info(
      status == "blocked" ? "user_blocked" : "user_unblocked",
      "success",
      status == "blocked" ? "User account blocked" : "User account unblocked",
      actor: caller.username,
      resource_type: "user",
      resource_id: username
    )
    row = find_account(username)
    Stackverse::Errors.not_found unless row
    render_json Stackverse::Serializers.user_account(row)
  end

  def audit_log
    require_role!("admin")
    page, size = query_params.page_and_size
    filters = {
      "actor" => "actor",
      "action" => "action",
      "targetType" => "target_type",
      "targetId" => "target_id"
    }
    conditions = [ "true" ]
    filters.each do |param_name, column|
      value = query_params.single(param_name)
      conditions << "#{column} = #{Stackverse::Sql.quote(value)}" if value
    end
    from = Stackverse::InputValidation.parse_datetime(query_params.single("from"), "from")
    to = Stackverse::InputValidation.parse_datetime(query_params.single("to"), "to")
    conditions << "created_at >= #{Stackverse::Sql.quote(from)}" if from
    conditions << "created_at <= #{Stackverse::Sql.quote(to)}" if to
    where_sql = conditions.join(" and ")
    rows = Stackverse::Sql.query(<<~SQL.squish)
      select * from audit_entries where #{where_sql}
      order by created_at desc, id desc
      limit #{size} offset #{page * size}
    SQL
    total = Stackverse::Sql.one("select count(*)::int as count from audit_entries where #{where_sql}")["count"].to_i
    render_json Stackverse::Sql.page(rows, page, size, total, Stackverse::Serializers.method(:audit))
  end

  def stats
    require_role!("moderator")
    today = Time.now.utc.to_date
    start = today - 29
    created_per_day = count_per_day("bookmarks", "created_at", start)
    active_per_day = count_per_day("user_accounts", "last_seen", start)
    daily = (0...30).map do |offset|
      day = (start + offset).iso8601
      {
        date: day,
        bookmarksCreated: created_per_day.fetch(day, 0),
        activeUsers: active_per_day.fetch(day, 0)
      }
    end
    totals = Stackverse::Sql.one(<<~SQL.squish)
      select
        (select count(*)::int from user_accounts) as users,
        (select count(*)::int from bookmarks) as bookmarks,
        (select count(*)::int from bookmarks where visibility = 'public') as public_bookmarks,
        (select count(*)::int from bookmarks where status = 'hidden') as hidden_bookmarks,
        (select count(*)::int from reports where status = 'open') as open_reports
    SQL
    payload = {
      totals: {
        users: totals["users"].to_i,
        bookmarks: totals["bookmarks"].to_i,
        publicBookmarks: totals["public_bookmarks"].to_i,
        hiddenBookmarks: totals["hidden_bookmarks"].to_i,
        openReports: totals["open_reports"].to_i
      },
      daily: daily,
      topTags: Stackverse::Sql.query(<<~SQL.squish)
        select tag, count(*)::int as count
        from bookmarks, unnest(tags) as tag
        group by tag
        order by count desc, tag asc
        limit 10
      SQL
    }
    render_etag payload
  end

  private

  def with_bookmark_count
    <<~SQL.squish
      select u.*, (select count(*)::int from bookmarks b where b.owner = u.username) as bookmark_count
      from user_accounts u
    SQL
  end

  def find_account(username)
    Stackverse::Sql.one("#{with_bookmark_count} where u.username = #{Stackverse::Sql.quote(username)}")
  end

  def reopen_report!(report, caller)
    conflict = Stackverse::Sql.one(<<~SQL.squish)
      select 1 from reports
      where bookmark_id = #{Stackverse::Sql.quote(report["bookmark_id"])}::uuid
        and reporter = #{Stackverse::Sql.quote(report["reporter"])}
        and status = 'open'
        and id <> #{Stackverse::Sql.quote(report["id"])}::uuid
    SQL
    Stackverse::Errors.conflict("The reporter already has another open report on this bookmark.") if conflict
    row = Stackverse::Sql.one(<<~SQL.squish)
      update reports
      set status = 'open', resolved_by = null, resolved_at = null, resolution_note = null
      where id = #{Stackverse::Sql.quote(report["id"])}::uuid
      returning *
    SQL
    Stackverse::Audit.record(caller.username, "report.reopened", "report", report["id"].to_s, { bookmarkId: report["bookmark_id"].to_s })
    Stackverse::EventLog.info(
      "report_reopened",
      "success",
      "Report re-opened",
      actor: caller.username,
      resource_type: "report",
      resource_id: report["id"].to_s,
      bookmark_id: report["bookmark_id"].to_s
    )
    row
  end

  def resolve_one!(report, resolution, actor, note, auto_resolved)
    row = Stackverse::Sql.one(<<~SQL.squish)
      update reports
      set status = #{Stackverse::Sql.quote(resolution)},
          resolved_by = #{Stackverse::Sql.quote(actor)},
          resolved_at = #{Stackverse::Sql.quote(Stackverse::Clock.now)},
          resolution_note = #{Stackverse::Sql.quote(note)}
      where id = #{Stackverse::Sql.quote(report["id"])}::uuid
      returning *
    SQL
    Stackverse::Audit.record(
      actor,
      "report.resolved",
      "report",
      report["id"].to_s,
      {
        bookmarkId: report["bookmark_id"].to_s,
        resolution: resolution,
        note: note,
        autoResolved: auto_resolved
      }
    )
    Stackverse::EventLog.info(
      "report_resolved",
      "success",
      "Report resolved",
      actor: actor,
      resource_type: "report",
      resource_id: report["id"].to_s,
      bookmark_id: report["bookmark_id"].to_s,
      resolution: resolution,
      auto_resolved: auto_resolved
    )
    row
  end

  def hide_bookmark!(bookmark_id, actor, note)
    bookmark = Stackverse::Sql.one("select * from bookmarks where id = #{Stackverse::Sql.quote(bookmark_id)}::uuid")
    Stackverse::Errors.not_found unless bookmark
    return if bookmark["status"] == "hidden"

    Stackverse::Sql.connection.execute(<<~SQL.squish)
      update bookmarks
      set status = 'hidden', updated_at = #{Stackverse::Sql.quote(Stackverse::Clock.now)}
      where id = #{Stackverse::Sql.quote(bookmark_id)}::uuid
    SQL
    Stackverse::Audit.record(
      actor,
      "bookmark.status-changed",
      "bookmark",
      bookmark_id,
      { from: "active", to: "hidden", note: note }
    )
    Stackverse::EventLog.info(
      "bookmark_status_changed",
      "success",
      "Bookmark hidden by an actioned report",
      actor: actor,
      resource_type: "bookmark",
      resource_id: bookmark_id,
      from: "active",
      to: "hidden"
    )
  end

  def count_per_day(table, column, start_date)
    allowed = {
      [ "bookmarks", "created_at" ] => true,
      [ "user_accounts", "last_seen" ] => true
    }
    raise ArgumentError, "unsupported daily series" unless allowed[[ table, column ]]

    start_time = start_date.to_time.utc
    rows = Stackverse::Sql.query(<<~SQL.squish)
      select (#{column} at time zone 'UTC')::date::text as day, count(*)::int as count
      from #{table}
      where #{column} >= #{Stackverse::Sql.quote(start_time)}
      group by day
    SQL
    rows.each_with_object({}) do |row, counts|
      counts[row["day"]] = row["count"].to_i
    end
  end
end
